qiita_db/util.py

```diff
@@ -604,107 +617,128 @@ def insert_filepaths(filepaths, obj_id, table, filepath_table, conn_handler,
+                 RETURNING filepath_id""".format(filepath_table)
+        idx = TRN.index
+        TRN.add(sql, values, many=True)
+        # Since we added the query with many=True, we've added len(values)
+        # queries to the transaction, so the ids are in the last idx queries
+        return list(chain.from_iterable(
+            chain.from_iterable(TRN.execute()[idx:])))
 
 
 def purge_filepaths():
     r"""Goes over the filepath table and remove all the filepaths that are not
-    used in any place"""
-    conn_handler = SQLConnectionHandler()
-
-    # Get all the (table, column) pairs that reference to the filepath table
-    # Code adapted from http://stackoverflow.com/q/5347050/3746629
-    table_cols_pairs = conn_handler.execute_fetchall(
-        """SELECT R.TABLE_NAME, R.column_name
-        FROM INFORMATION_SCHEMA.CONSTRAINT_COLUMN_USAGE u
-        INNER JOIN INFORMATION_SCHEMA.REFERENTIAL_CONSTRAINTS FK
-            ON U.CONSTRAINT_CATALOG = FK.UNIQUE_CONSTRAINT_CATALOG
-            AND U.CONSTRAINT_SCHEMA = FK.UNIQUE_CONSTRAINT_SCHEMA
-            AND U.CONSTRAINT_NAME = FK.UNIQUE_CONSTRAINT_NAME
-        INNER JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE R
-            ON R.CONSTRAINT_CATALOG = FK.CONSTRAINT_CATALOG
-            AND R.CONSTRAINT_SCHEMA = FK.CONSTRAINT_SCHEMA
-            AND R.CONSTRAINT_NAME = FK.CONSTRAINT_NAME
-        WHERE U.COLUMN_NAME = 'filepath_id'
-            AND U.TABLE_SCHEMA = 'qiita'
-            AND U.TABLE_NAME = 'filepath'""")
-
-    union_str = " UNION ".join(
-        ["SELECT %s FROM qiita.%s WHERE %s IS NOT NULL" % (col, table, col)
-         for table, col in table_cols_pairs])
-    # Get all the filepaths from the filepath table that are not
-    # referenced from any place in the database
-    fps = conn_handler.execute_fetchall(
-        """SELECT filepath_id, filepath, filepath_type, data_directory_id
-        FROM qiita.filepath FP JOIN qiita.filepath_type FPT
-            ON FP.filepath_type_id = FPT.filepath_type_id
-        WHERE filepath_id NOT IN (%s)""" % union_str)
-
-    # We can now go over and remove all the filepaths
-    for fp_id, fp, fp_type, dd_id in fps:
-        conn_handler.execute("DELETE FROM qiita.filepath WHERE filepath_id=%s",
-                             (fp_id,))
-
-        # Remove the data
-        fp = join(get_mountpoint_path_by_id(dd_id), fp)
-        if exists(fp):
-            if fp_type is 'directory':
-                rmtree(fp)
-            else:
-                remove(fp)
+    used in any place
+
+    Raises
+    ------
+    IOError
+        If any error occurs while removing the fileapths from the filesystem
+
+    Notes
+    -----
+    This function can potentially leave the DB and the filesystem out of
+    sync if purge_filepaths is execute inside a bigger transaction. Thus,
+    care should be taken when using this function and do not include it in
+    a bigger transactions, as it can leave the database pointing to files that
```

Comment:
I am working right now on a way of adding something to the Transaction object that will make sure that all files that are in the filepath table are in the filesystem.

However, consistency in the other way around is not ensured, i.e. files that are present in the FS are not in the DB. For example, in purge_filepaths these files have been removed from the DB but for some error they're not removed from the FS.